### PR TITLE
Issue 83 OAuth flow test 

### PR DESCRIPTION
**Pull request by finyak4**
## Changes
- Added custom pipeline for is_active field
- Overwritten user_details pipeline to prevent first and last name overwriting with empty values
- Changes test due to changes in code
- Polished settings for social_django

*1.1. Google with verified email, first and last names*

*1.2. GitHub without first and last names but not overwritten with empty strings*

*2.1. GitHub with verified email and without names*

*2.2. Names are overwritten with new, not empty values*

### CODE REVIEW

File: users/pipelines.py [hunk 1]
```diff
@@ -25,9 +25,59 @@ def create_or_update_user(strategy, details, backend, user=None, *args, **kwargs
...
+                emails = social.extra_data.get('emails', []) 
+                user_email = user.email or details.get('email', '')
+                for email_data in emails:
+                    if (isinstance(email_data, dict) and 
+                        email_data.get('email') == user_email and 
+                        email_data.get('verified', False)):
+                        verified_email = True
+                        break
+            else:
+                verified_email = True
```
Comment (mehalyna) [inline users/pipelines.py:55]: The else block for Github verification defaults to trusted if emails are missing, which could be insecure. Require explicit email verification for Github and handle missing emails defensively.
```
else:
    verified_email = False  # do not trust if emails are missing
```

File: users/pipelines.py [hunk 2]
```diff
@@ -25,9 +25,59 @@ def create_or_update_user(strategy, details, backend, user=None, *args, **kwargs
...
+            user.is_active = True
+            user.save()
+
+def safe_user_details(strategy, details, backend, user=None, *args, **kwargs):
+    """
+    Update user with provider details, but never touch first_name/last_name.
+    Returns {'user': user} to keep the pipeline happy.
+    """
+    if user is None:
+        return
```
Comment (mehalyna) [inline users/pipelines.py:66]: The function returns None if user is None, potentially breaking the pipeline contract. Return {'user': None} to maintain pipeline consistency.
```
if user is None:
    return {'user': None}
```

File: users/pipelines.py [hunk 3]
```diff
@@ -25,9 +25,59 @@ def create_or_update_user(strategy, details, backend, user=None, *args, **kwargs
...
+
+    changed = False
+    for field, value in details.items():
+        if field in ['first_name', 'last_name']:
+            continue
+        if not value:
+            continue
```
Comment (mehalyna) [inline users/pipelines.py:73]: The check if not value will skip falsy but valid values such as empty strings or False. Check explicitly for None instead to avoid skipping valid updates.
```
if value is None:
    continue
```

File: users/pipelines.py [hunk 4]
```diff
@@ -25,9 +25,59 @@ def create_or_update_user(strategy, details, backend, user=None, *args, **kwargs
             last_name=details.get('last_name', ''),
             role=UserRole.objects.get(role='user')
         )
-        send_welcome_oauth_email_task.delay(
-            subject="Welcome!",
-            message="Thanks for signing up via OAuth",
-            recipient_list=[user.email]
-        )
```
Comment (mehalyna) [inline users/pipelines.py:32]: The email sending logic was removed entirely, breaking welcome email notifications for new users. Restore the email notification for new users after creation.
```
if created:
    send_welcome_oauth_email_task.delay(
        subject="Welcome!",
        message="Thanks for signing up via OAuth",
        recipient_list=[user.email]
    )
```

File: users/views/oauth_view.py [hunk 5]
```diff
@@ -172,7 +161,7 @@ def send_welcome_email(self, user, provider):
         """
         PROVIDER_MAP = {"google": "Google", "github": "GitHub"}
         provider_name = PROVIDER_MAP.get(provider, provider)
-        action = "registered" if not user.last_login else "logged in"
+        action = "registered" if timezone.now() - user.created_at < timedelta(seconds=5) else "logged in"
```
Comment (mehalyna) [inline users/views/oauth_view.py:164]: The 5-second threshold for determining registration is arbitrary and could fail for delayed requests. Consider using a flag or more robust method to detect new registrations.
```
# Example: Add a boolean `is_new_user` at creation and use it here
action = "registered" if is_new_user else "logged in"
```

**Comment by mehalyna** [comment]
The job failed due to multiple database integrity errors and unhandled edge cases in test or fixture setup. The most critical errors are:

1. **IntegrityError: duplicate key value violates unique constraint**  
   - Example:  
     ```
     ERROR:  duplicate key value violates unique constraint "unique_investor_project"
     DETAIL:  Key (investor_id, project_id)=(3, 6) already exists.
     STATEMENT:  INSERT INTO "subscriptions" ...
     ```
   - This means your tests or setup code attempts to insert records with (investor_id, project_id) pairs (and other keys, such as categories.name and saved_startups) that already exist.

2. **OperationalError: database "test_test_db" is being accessed by other users**  
   - Example:  
     ```
     OperationalError: database "test_test_db" is being accessed by other users
     DETAIL:  There is 1 other session using the database.
     ```
   - This often occurs if a test or fixture doesn't close DB connections or if parallel tests are not isolated.

---

## Solution Steps

### 1. Fix Unique Constraint Violations in Test Setup

**Problem:**  
Tests are inserting duplicate values for unique fields.

**Solution:**  
- Review your test setup (`setUp`, fixtures, or factory calls).  
- Before creating a new record, check if one already exists, or use `.get_or_create`, or ensure unique data for each test run.

**Example Fix:**
```python
# For subscriptions
Subscription.objects.get_or_create(
    investor_id=3, project_id=6,
    defaults={'amount': 70, 'investment_share': 0.00}
)

# For categories
Category.objects.get_or_create(
    name='Health',
    defaults={'description': '', 'created_at': timezone.now()}
)

# For saved_startups
SavedStartup.objects.get_or_create(
    investor_profile_id=1, startup_profile_id=1,
    defaults={
        'saved_at': timezone.now(),
        'status': 'watching',
        # ... other fields ...
    }
)
```

### 2. Ensure Proper Test Teardown and Isolation

**Problem:**  
Database is being accessed by other sessions, causing teardown to fail.

**Solution:**  
- Make sure all DB connections are closed after each test.
- If using pytest, add `--reuse-db` or `--no-db` as needed.
- Avoid parallel test execution if the DB isn't properly isolated.
- Explicitly close or rollback any open transactions in your test code.

**Example Fix:**
```python
from django.db import connection

def tearDown(self):
    super().tearDown()
    connection.close()
```

### 3. General Recommendations

- If you use test factories (FactoryBoy, etc.), set `django_get_or_create = ('unique_field',)` for unique fields.
- Do not hardcode IDs; use auto-generated ones or query for existing objects.
- Add logging in your test setup to verify when/where duplicates are created.

---

## Next Steps

- Update your test setup to avoid duplicate entries for unique constraints.
- Ensure test teardown closes all DB connections.
- Rerun the job after applying these fixes.

File: users/pipelines.py (hunk 4 above)
Comment (finyak4) [reply users/pipelines.py]: Restored

File: users/views/oauth_view.py (hunk 5 above)
Comment (finyak4) [reply users/views/oauth_view.py]: Changed

**Review by Raidriar28 — APPROVED** [review]
Good job!
